Archive/Examples/IfNormalization/Result.lean

```diff
@@ -0,0 +1,115 @@
+/-
+Copyright (c) 2023 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+-/
+import Archive.Examples.IfNormalization.Statement
+import Mathlib.Data.List.AList
+import Mathlib.Tactic.Recall
+
+/-!
+# A solution to the if normalization challenge in Lean.
+
+See `Statement.lean` for background.
+-/
+
+set_option autoImplicit true
+
+macro "◾" : tactic => `(tactic| aesop)
+macro "◾" : term => `(term| by aesop)
+
+namespace IfExpr
+
+/-!
+We add some local simp lemmas so we can unfold the definitions of the normalization condition.
+-/
+attribute [local simp] normalized hasNestedIf hasConstantIf hasRedundantIf disjoint vars
+  List.disjoint
+
+/-!
+Adding these lemmas to the simp set allows Lean to handle the termination proof automatically.
+-/
+attribute [local simp] Nat.lt_add_one_iff le_add_of_le_right
+
+/-!
+Some further simp lemmas for handling if-then-else statements.
+-/
+attribute [local simp] apply_ite ite_eq_iff'
+
+
+/-!
+Simp lemmas for `eval`.
+We don't want a `simp` lemma for `(ite i t e).eval` in general, only once we know the shape of `i`.
+-/
+@[simp] theorem eval_lit : (lit b).eval f  = b := rfl
+@[simp] theorem eval_var : (var i).eval f  = f i := rfl
+@[simp] theorem eval_ite_lit :
+    (ite (.lit b) t e).eval f = bif b then t.eval f else e.eval f := rfl
+@[simp] theorem eval_ite_var :
+    (ite (.var i) t e).eval f = bif f i then t.eval f else e.eval f := rfl
+@[simp] theorem eval_ite_ite :
+    (ite (ite a b c) d e).eval f = (ite a (ite b d e) (ite c d e)).eval f := by
+  cases h : eval f a <;> simp_all [eval]
+
+/-- Custom size function for if-expressions, used for proving termination. -/
+@[simp] def normSize : IfExpr → Nat
+  | lit _ => 0
+  | var _ => 1
+  | .ite i t e => 2 * normSize i + max (normSize t) (normSize e) + 1
+
+/-- Normalizes the expression at the same time as assigning all variables in
+`l` to the literal booleans given by `l` -/
+def normalize (l : AList (fun _ : ℕ => Bool)) :
+    (e : IfExpr) → { e' : IfExpr //
+        (∀ f, e'.eval f = e.eval (fun w => (l.lookup w).elim (f w) (fun b => b)))
+        ∧ e'.normalized
+        ∧ ∀ (v : ℕ), v ∈ vars e' → l.lookup v = none }
+  | lit b => ⟨lit b, ◾⟩
+  | var v =>
+    match h : l.lookup v with
+    | none => ⟨var v, ◾⟩
+    | some b => ⟨lit b, ◾⟩
+  | .ite (lit true)   t e => have := normalize l t; ◾
+  | .ite (lit false)  t e => have := normalize l e; ◾
+  | .ite (.ite a b c) t e => have := normalize l (.ite a (.ite b t e) (.ite c t e)); ◾
```

Comment:
You're using a tactic to write a program. This is interesting.

Reply:
Do you think it's overkill? :-)

Reply:
I think it probably is. It makes it harder to see what the program actually does, and to understand performance considerations. You also want to make sure it's not doing something weird. I don't think it is, but it's easier to make sure of that fact if you actually write explicitly what's supposed to happen.

Reply:
Okay, I've written the data explicitly again.

It would be nice to have something even shorter than `⟨t'.1, ◾⟩` when you want to specify some data and then let `aesop` discharge the proofs.

Reply:
We could define `Subtype.of (val : \a) (property : p val := by aesop)`, and then just be able to write `of t'.1`.

It's a pity we can't have coercions that depend on tactics. :-)